File: EKMC/EKMC/kMC_algorithm_methods/get_rate_constant_methods/get_Marcus_rate_constants_data.py

import os

from time import time
from datetime import timedelta
from itertools import count
from math import exp, log
from numpy.random import normal, uniform
from random import choice, choices

def get_Marcus_rate_constants_data(current_molecule_description, all_local_neighbourhoods, energetic_disorder, coupling_disorder, graph):
	current_molecule_deltaE_disorder, neighbouring_molecule_descriptions, rate_constants = get_rate_constants_to_connecting_molecules(current_molecule_description, all_local_neighbourhoods, energetic_disorder, coupling_disorder, graph)
	return current_molecule_deltaE_disorder, neighbouring_molecule_descriptions, rate_constants

def get_rate_constants_to_connecting_molecules(current_molecule_description, all_local_neighbourhoods, energetic_disorder, coupling_disorder_variable, graph):
	"""
	This method is designed to gather the rate constants between the current molecule in the current cell point and every other neighbouring molecule in surrounding unit cells within the rCut_neighbourhood radius. 

	Parameters
	----------
	current_molecule_name : int
		This is the molecule that the exciton is currently on.
	current_cell_point : tuple
		This is the cell that the exciton is currently in.
	all_local_neighbourhoods : dict.
		This dictionary contains all the information about the neighbourhoods that surrounded each molecule in your crystal.
	energetic_disorder : float
		This value indicates the disorder given in the delta E value.
	coupling_disorder_variable : float
		This value indicates the disorder given in the V (coupling) value.
	graph : kMC_graph
		This graph holds all the information about the molecules in the various positions in the crystal.

	Returns
	-------
	other_molecule_descriptions : list
		This is the name and the cell_point that the the exciton can move to from the current molecule.
	other_molecule_deltaE_disorders : list
		These are the deltaE values for the other molecules that are kinetically attached to the current molecule. 
	rate_constants : list
		These are the rate constants for the exciton to move from the current molecule to another molecule that maybe in another unit cell.
	edges_details : dict.
		These are other edge details for the possible routes for the exciton to move from the current molecule to another molecule that maybe in another unit cell.
	"""

	# First, get the molecule current name and current cell point from the current_molecule_description tuple
	current_molecule_name, current_cell_point = current_molecule_description

	# Second, initalise the list and dictionaries to record data into
	neighbouring_molecule_descriptions = []
	rate_constants = []
	edges_details = {}

	# Third, get the energy for this molecule that has had disorder applied to it.
	current_molecule_E_with_disorder = graph.get_E_with_disorder(current_molecule_description)
	if current_molecule_E_with_disorder is None:
		current_molecule_E_with_disorder = normal(0.0,energetic_disorder) # Set energy of molecule to 0.0 eV, and get random normal distributed number around 0.0 eV with width of energetic_disorder (given in eV)
		graph.add_E_with_disorder(current_molecule_description, current_molecule_E_with_disorder)

	# Fourth, obtain the relative local neighbourhood for the current molecule
	local_neighbourhoods = all_local_neighbourhoods[current_molecule_name]

	# Fifth, obtain all the rate constants and data for an exciton moving from the current molecule to another molecule that maybe in another unit cell.
	for neighbouring_molecule_name in local_neighbourhoods.keys():
		for relative_cell_point, (coupling_energy, M_constant, X_constant, Y_constant, Z_constant) in local_neighbourhoods[neighbouring_molecule_name].items():

			# 5.1: Get the exist cell_point for the other molecule from current_cell_point + relative_cell_point
			neighbouring_cell_point = tuple(a+b for a, b in zip(current_cell_point, relative_cell_point))
			neighbouring_molecule_description = (neighbouring_molecule_name, neighbouring_cell_point)

			# 5.2.1: Get the E value based on disorder for this other molecule.
			other_molecule_E_with_disorder = graph.get_E_with_disorder(neighbouring_molecule_description)
			if other_molecule_E_with_disorder is None:
				other_molecule_E_with_disorder = normal(0.0,energetic_disorder) # Set energy of molecule to 0.0 eV, and get random normal distributed number around 0.0 eV with width of energetic_disorder (given in eV)
				graph.add_E_with_disorder(neighbouring_molecule_description, other_molecule_E_with_disorder)

			# 5.2.2: Obtain the randomly generated number to describe the energetic and coupling disorders, based on a normal distribution. 
			V_with_disorder = graph.get_V_with_disorder(current_molecule_description, neighbouring_molecule_description)
			if V_with_disorder is None:

				# 5.2.2.1: if the coupling_disorder is a string and ends with %, take the coupling_disorder as the percentage of the dimers coupling energy
				if isinstance(coupling_disorder_variable,str) and coupling_disorder_variable.endswith('%'):
					coupling_disorder_percentage = float(coupling_disorder_variable.replace('%',''))
					coupling_disorder = abs(coupling_energy * (coupling_disorder_percentage/100.0))
				else:
					coupling_disorder = coupling_disorder_variable
				
				# 5.2.2.2: Obtain the coupling disorder for the disorder energy. coupling_energy and coupling_disorder given in eV. 
				V_with_disorder = normal(coupling_energy, coupling_disorder)

				# 5.2.2.3: Add this V_with_disorder to the KMC crystal database graph 
				graph.add_V_with_disorder(current_molecule_description, neighbouring_molecule_description, V_with_disorder)

			# 5.3: get the difference in energy between the two molecules.
			deltaE_with_disorders = (other_molecule_E_with_disorder - current_molecule_E_with_disorder)

			# 5.4: Obtain the rate constant for the exciton to move from the current molecule to another molecule that maybe in another unit cell.
			k_12 = (V_with_disorder**2.0)*M_constant*exp(-((deltaE_with_disorders**2.0)*X_constant + deltaE_with_disorders*Y_constant + Z_constant))

			# 5.5: Add the rate constant data to the storage list and dictionaries. 
			neighbouring_molecule_descriptions.append(neighbouring_molecule_description)
			rate_constants.append(k_12)
			
	# Sixth, return current_molecule_E_with_disorder, neighbouring_molecule_descriptions, and rate_constants
	return current_molecule_E_with_disorder, neighbouring_molecule_descriptions, rate_constants

File: EKMC/EKMC/KMC_algorithm/Running_KMC_Methods/Rate_Constant_Methods/get_marcus_rate_constants_data.cpp
/**
 * get_marcus_rate_constants_data.cpp, Geoffrey Weal, 30/5/23
 * 
 * This algorithm is designed to obtain the exciton hopping rate constants between molecule in a crystal in accordance to Marcus Theory. 
 */
#include <iostream>
#include <list>
#include <vector>
#include <cstdlib>
#include <cmath>
#include <unordered_map>
using namespace std;
#include "get_E_with_disorder.h"
#include "get_V_with_disorder.h"
#include "get_marcus_rate_constants_data.h"

tuple<long double, list<tuple<int,int,int,int>>, list<long double>> get_marcus_rate_constants_data(int current_molecule_name, int* current_cell_point, 
	long double M_constant, long double X_constant, long double energetic_disorder_value, bool energetic_disorder_is_percent, 
	long double coupling_disorder_value, bool coupling_disorder_is_percent, unordered_map<int, long double>* molecule_bandgap_energies, 
	unordered_map<tuple<int,int>, long double, hash_tuple_DRE>* dimer_reorganisation_energies,
	unordered_map<int, vector<tuple<int,int,int,int,long double>>>* coupling_value_data,
	Molecule_Energetic_Disorder_Database* molecule_energetic_disorder_database, Rate_Constant_Database* rate_constant_database) {
	/**
	 * This algorithm is designed to obtain the exciton hopping rate constants between molecule in a crystal in accordance to Marcus Theory. 
	 * 
	 * @param current_molecule_name This is the molecule that the exciton is currently on.
	 * @param current_cell_point This is the cell that the exciton is currently in.
	 * @param M_constant This is the M constant in the Marcus Theory Rate law. This is a constant for every dimer in this crystal.
	 * @param X_constant This is the X constant in the Marcus Theory Rate law. This is a constant for every dimer in this crystal.
	 * @param energetic_disorder_value This is the energetic (bandgap) disorder value, either given as a standard deviation (in eV), or as a percentage of a energy (bandgap) for a molecule. 
	 * @param energetic_disorder_is_percent If True, energetic_disorder_value is a percentage. If False, energetic_disorder_value is a standard deviation (in eV).
	 * @param coupling_disorder_value This is the coupling disorder value, either given as a standard deviation (in eV), or as a percentage of a coupling value for a dimer. 
	 * @param coupling_disorder_is_percent If True, coupling_disorder_value is a percentage. If False, coupling_disorder_value is a standard deviation (in eV).
	 * @param molecule_bandgap_energies This contains all the bandgap energies for each molecule in the crystal.
	 * @param dimer_reorganisation_energies This contains all the reorganisation energies for each dimer in the crystal. 
	 * @param coupling_value_data This unordered_map contains all the information about the neighbourhoods that surrounded each molecule in your crystal, including coupling values for each dimer pair.
	 * @param molecule_energetic_disorder_database This map holds all the energies (bandgap) for each molecule sampled in a KMC simulation. 
	 * @param rate_constant_database This map holds all the rate constants for each dimer sampled in a KMC simulation. 
	 * 
	 * @returns current_molecule_donor_E_with_disorder: The energy of the current molecule the exciton is on, including disorder (in eV); neighbouring_molecule_descriptions: The molecules and the absolute unit cell positions of neighbour molecules that are coupled to the current molecule; rate_constants: The exciton hopping rate constants for an exciton hopping from the current molecule to the neighbouring molecules about it that it is coupled to.
	 */

	// First, initalise the list and dictionaries to record data into
	list<tuple<int,int,int,int>> neighbouring_molecule_descriptions;
	list<long double> rate_constants;

	// Second, get the energy for this molecule that has had disorder applied to it.
	long double current_molecule_donor_E_with_disorder = get_E_with_disorder(current_molecule_name, current_cell_point, molecule_energetic_disorder_database, molecule_bandgap_energies, energetic_disorder_value, energetic_disorder_is_percent);

	// Third, obtain the relative local neighbourhood for the current molecule
	vector<tuple<int,int,int,int,long double>>* local_coupling_value_data = &(*coupling_value_data)[current_molecule_name];

	// Fourth, obtain all the rate constants and data for an exciton moving from the current molecule to another molecule that maybe in another unit cell.
	for (const auto& local_neighbourhood : (*local_coupling_value_data)){

		// 4.1: Obtain the neighbouring molecule name.
		int neighbouring_molecule_name = get<0>(local_neighbourhood);

		// 4.2: Obtain the absolute position of the potential acceptor molecule by 
		//      adding the absolute position of the donor molecule to the relative 
		//      unit cell displacement of molecule 2 to molecule 1.
		int neighbouring_cell_point[3] = {get<1>(local_neighbourhood), get<2>(local_neighbourhood), get<3>(local_neighbourhood)};
		for (int index = 0; index < 3; index++) {
			neighbouring_cell_point[index] = neighbouring_cell_point[index] + current_cell_point[index];
		}

		// 4.3: Obtain the rate constant search key for looking through the rate_constant_database.
		tuple<int,int,int,int, int,int,int,int> R_search_key = make_tuple(current_molecule_name, current_cell_point[0], current_cell_point[1], current_cell_point[2], neighbouring_molecule_name, neighbouring_cell_point[0], neighbouring_cell_point[1], neighbouring_cell_point[2]);

		// 4.4: obtain the rate constant for this dimer in the crystal.
		long double k_12;
		if (! rate_constant_database->contains(R_search_key)) {

			// 4.4.1: Obtain the energy for the neighbouring (acceptor) molecule that has had disorder applied to it.
			long double neighbouring_molecule_acceptor_E_with_disorder = get_E_with_disorder(neighbouring_molecule_name, neighbouring_cell_point, molecule_energetic_disorder_database, molecule_bandgap_energies, energetic_disorder_value, energetic_disorder_is_percent);

			// 4.4.2: Obtain the deltaE for this exciton hop with included disorders.
			long double deltaE_with_disorders = neighbouring_molecule_acceptor_E_with_disorder - current_molecule_donor_E_with_disorder;

			// 4.4.3: Obtain the coupling between current_molecule_name and neighbouring_molecule_name at relative unit cell displacement neighbouring_cell_point
			long double coupling_value = get<4>(local_neighbourhood);

			// 4.4.4: Obtain the randomly generated number to describe the energetic and coupling disorders, based on a normal distribution. 
			long double V_with_disorder = get_V_with_disorder(coupling_value, coupling_disorder_value, coupling_disorder_is_percent);

			// 4.4.5: Obtain the reorganisation energy for the exciton moving from current molecule (in the excited geometry structure) to the neighbouring molecule (in the ground geometry structure).
			long double reorganisation_energy = (*dimer_reorganisation_energies)[make_tuple(current_molecule_name,neighbouring_molecule_name)];

			// 4.4.6: Obtain the rate constant for the exciton to move from the current molecule to another molecule that maybe in another unit cell.
			long double prefix_value = pow(abs(V_with_disorder),2.0) / pow(reorganisation_energy,0.5);
			long double exp_value = pow(deltaE_with_disorders + reorganisation_energy,2.0) / reorganisation_energy;
			k_12 = prefix_value * M_constant * exp( -X_constant * exp_value );

			// 4.4.7: Add k_12 to rate_constant_database with key R_search_key
			rate_constant_database->add(R_search_key, k_12);

		} else {

			// 4.4.8: Get the k_12 from rate_constant_database for R_search_key key.
			k_12 = rate_constant_database->get(R_search_key);
		}

		// 4.5: Add the rate constant data to the storage list and dictionaries. 
		neighbouring_molecule_descriptions.push_back(make_tuple(neighbouring_molecule_name, neighbouring_cell_point[0], neighbouring_cell_point[1], neighbouring_cell_point[2]));
		rate_constants.push_back(k_12);

	}

    // Fifth, return current_molecule_donor_E_with_disorder, neighbouring_molecule_descriptions, and rate_constants.
	return make_tuple(current_molecule_donor_E_with_disorder, neighbouring_molecule_descriptions, rate_constants);

}
